### ISSUE TITLE
k3s: package all supported versions

### ISSUE THREAD

**Issue by yajo** [NixOS/nixpkgs#213943]
### Describe the bug
K3s, just like with normal k8s, has a very specific [version skew policy](https://kubernetes.io/releases/version-skew-policy/#supported-component-upgrade-order).

Summarizing, to have a successful HA cluster upgrade, you need to:

1. Upgrade your server nodes to the latest *patch* version available. One node at a time.
2. Do the same operation with worker nodes, in no particular order.
3. Repeat steps 1-2, but this time upgrading to the latest patch version in the *next minor* version.
4. Repeat step 3 until reached the *desired minor* version.

Latest k3s in NixOS 22.05: https://github.com/NixOS/nixpkgs/blob/0874168639713f547c05947c76124f78441ea46c/pkgs/applications/networking/cluster/k3s/default.nix#L49

And in Nixos 22.11: https://github.com/NixOS/nixpkgs/blob/0218941ea68b4c625533bead7bbb94ccce52dceb/pkgs/applications/networking/cluster/k3s/default.nix#L50

**There's no 1.24.x version available** AFAICS.

This problem unnecessarily complicates the upgrade from NixOS 22.05 to 21.11. 

### Expected behavior
A similar problem happens with python, and nixpkgs has `python38`, `python39`, `python310` and `python311` packages. Of course, there's a default rolling `python3` one too.

I would expect to be able to have the same structure for k3s. There should be `k3s123`, `k3s124`, `k3s125`, `k3s126`... and a `k3s` one which points to the latest release always.

That would allow decoupling k3s from NixOS upgrades. It would also let us keep on getting patch updates for older k3s versions in newer NixOS systems (which BTW is a requirement for a safe upgrade, as indicated above).

Given k8s clusters leverage critical production deployments usually, and upgrades involve drain of nodes and a lot of specific processes, I think this enhancement is very important and would make NixOS an even better candidate for a base system for these kind of workloads.

### Additional context
See [this related forum question](https://discourse.nixos.org/t/recommended-way-to-upgrade-k3s/24103).

In [this announcement](https://kubernetes.io/blog/2021/07/20/new-kubernetes-release-cadence/) and [this page](https://endoflife.date/kubernetes) you can read that the release cadence of kubernetes is about 3 minor versions per year. NixOS' has 2 stable versions per year. This means that, even if the k3s package is up to date all the time, **the same problem will happen in every stable NixOS release**.

### Notify maintainers

@euank @mic92 @superherointj

@moduon MT-1718

**Comment by yajo** [comment]
Looks great!

**Comment by superherointj** [comment]
> Looks great!

For a start, I can branch versions as:
- older versions that already exist in nixpkgs and works
- update script and tests disabled for older versions
- will keep update script and tests as it is now for 1_26 (will be rolling releases)

Update script will need to be updated, to accommodate new folder structure for new versions and see edge cases.

Should I add you as a K3s maintainer?

Note: I might help get this going but I have no interest in stable releases. I don't use stable at all. Just unstable.

**Comment by euank** [comment]
I'll chime in to say that I'm completely in favor of having all the supported k3s versions present in nixpkgs, which I believe is typically 3 versions.
This is something I've wanted for myself in the past as well quite a few times.

I don't think we should keep k3s releases in unstable if they're no longer supported by upstream (at that point, if someone wants them, they can use an older nixpkgs or grab em themselves), but for as long as kubernetes / rancher may be releasing new updates for the version, I think keeping it around will make sense.

We'll also want to backport to stable releases appropriate patch releases (any where the corresponding minor version is in that stable release).

I think the most complexity in this will be figuring out the best way to structure the package internally, for maintainability, and the best way to automate those changes to unstable.

**Comment by yajo** [comment]
After #214026 is merged, I think we should structure it like this:

- For older packages, the copy-paste done there is just enough.
- When 1.27 comes out, we could try to "abstract" its derivation with 1.26, to keep it DRY.
- When 1.23-1.25 fade away, just remove them.
- If somebody still wants to update 1.23-1.25 before they lose upstream support, they can send a separate PR as usual. FWIW I'm planning to do it.

It seems a simple approach to keep the older derivations available, while focusing maintenance efforts in newer ones.
